pkg/client/client.go

```diff
@@ -352,6 +387,188 @@ func (c *serviceImpl) doRequest(method, path string, reqpl, respl any) error {
 	return nil
 }
 
+const openConnections = 10
+
+type fileInfo struct {
+	Name        string
+	TotalChunks int
+}
+
+type filesUploader struct {
+	Client         *http.Client
+	EndpointURL    string
+	Filenames      []string
+	ChunkSizeBytes int64
+	DumpOut        io.Writer
+}
+
+func (u *filesUploader) Upload() error {
+	infos, err := u.getFilesInfos()
+	if err != nil {
+		return err
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	jobsChan := make(chan uploadChunkJob)
+	resultsChan := u.startWorkers(ctx, jobsChan)
+	go func() {
+		defer close(jobsChan)
+		for _, info := range infos {
+			for i := 0; i < info.TotalChunks; i++ {
+				select {
+				case <-ctx.Done():
+					return
+				default:
+					j := uploadChunkJob{
+						Filename:       info.Name,
+						ChunkNumber:    i + 1,
+						TotalChunks:    info.TotalChunks,
+						ChunkSizeBytes: u.ChunkSizeBytes,
+					}
+					jobsChan <- j
+				}
+			}
+		}
+	}()
+	// Only first error will be returned.
+	var returnErr error
+	for err := range resultsChan {
+		if returnErr != nil {
+			continue
+		}
+		if err != nil {
+			returnErr = err
+			cancel()
+			// Do not return from here and let the cancellation logic to propagate, resultsChan
+			// will be closed eventually.
+		}
+	}
+	return returnErr
+}
+
+func (u *filesUploader) getFilesInfos() ([]fileInfo, error) {
+	var infos []fileInfo
+	for _, name := range u.Filenames {
+		stat, err := os.Stat(name)
+		if err != nil {
+			return nil, err
+		}
+		info := fileInfo{
+			Name:        name,
+			TotalChunks: int((stat.Size() + u.ChunkSizeBytes - 1) / u.ChunkSizeBytes),
+		}
+		infos = append(infos, info)
+	}
+	return infos, nil
+}
+
+func (u *filesUploader) startWorkers(ctx context.Context, jobsChan <-chan uploadChunkJob) <-chan error {
+	agg := make(chan error)
+	wg := sync.WaitGroup{}
+	for i := 0; i < openConnections; i++ {
+		wg.Add(1)
+		w := uploadChunkWorker{
+			Client:      u.Client,
+			EndpointURL: u.EndpointURL,
+			DumpOut:     u.DumpOut,
+			JobsChan:    jobsChan,
+		}
+		go func(ch <-chan error) {
+			for err := range ch {
+				agg <- err
+			}
+			wg.Done()
+		}(w.Start())
```

Comment:
nit: I struggled to understand this block, consider rewriting as
```
go func() {
	var ch <-chan error = w.Start() // or just ch := w.Start()
	for err := range ch {
		agg <- err
	}
	wg.Done()
}()
```

Reply:
good point, done